addons/minio/template/base/install.sh

```diff
@@ -10,10 +10,290 @@ function minio_pre_init() {
+    kubectl get endpoints --no-headers -n "$namespace" "$service" -o custom-columns=:subsets |grep -q none
+}
+
+# disable_minio_svc disables the minio service by tweaking its service selector.
+function disable_minio_svc() {
+    kubectl patch svc -n $MINIO_NAMESPACE minio -p '{"spec":{"selector":{"app":"does-not-exist"}}}'
+}
+
+# enable_minio_svc sets the minio service selector back to its default.
+function enable_minio_svc() {
+    kubectl patch svc -n $MINIO_NAMESPACE minio -p '{"spec":{"selector":{"app":"minio"}}}'
+}
+
+# prepare_volumes_for_migration sets the retention policy of the volumes to "retain" for both minio and its migration deployment.
+function prepare_volumes_for_migration() {
+    # get the pv in use by the minio deployment.
+    local minio_pv
+    minio_pv=$(kubectl get pvc --no-headers -n $MINIO_NAMESPACE minio-pv-claim -o custom-columns=:.spec.volumeName)
+    if [ -z "$minio_pv" ]; then
+        return 1
+    fi
+
+    # get the pv in use by the fs migration deployment.
+    local migration_pv
+    migration_pv=$(kubectl get pvc --no-headers -n $MINIO_NAMESPACE minio-migrate-fs-backend-pv-claim -o custom-columns=:.spec.volumeName)
+    if [ -z "$migration_pv" ]; then
+        return 1
+    fi
+
+    # set both of the pv retention policies to 'retain'.
+    kubectl patch pv "$minio_pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+    kubectl patch pv "$migration_pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+}
+
+# migrate_fs_backend migrates a minio running with 'fs' backend into the new default backend type (xl). spawns a new
+# minio deployment then copies the object buckets from the old deployment into the new one. this function generates
+# unavailability of the original minio service.
+function migrate_fs_backend() {
+    # if minio is not running then we can't migrate its data. minio may be deployed by means of a statefulset as well
+    # but on this case the running pods are guaranteed to not leverage the fs backend.
+    local minio_replicas
+    minio_replicas=$(kubectl get --no-headers deploy minio -n $MINIO_NAMESPACE -o custom-columns=:.spec.replicas 2>/dev/null)
+    if [ -z "$minio_replicas" ] || [ "$minio_replicas" = "0" ]; then
+        return
+    fi
+
+    # if minio does not use fs backend we don't need to migrate it.
+    if ! minio_uses_fs_format ; then
+        return
+    fi
+
+    printf "${YELLOW}\n"
+    printf "The installer has detected that the cluster is running a version of minio backed by the now legacy FS format.\n"
+    printf "To be able to upgrade to the new Minio version a migration will be necessary. During this migration, Minio\n"
+    printf "will be unavailable.\n"
+    printf "\n"
+    printf "For further information please check https://github.com/minio/minio/releases/tag/RELEASE.2022-10-29T06-21-33Z\n"
+    printf "${NC}\n"
+    printf "Would you like to proceed with the migration ?"
+    if ! confirmN; then
+        bail "Not migrating"
+    fi
```

Comment:
Is it possible to move this and the check for minio_uses_fs_format to the minio_pre_init function? it is always better to provide feedback to the user and bail early.

Reply:
Yes, sounds good. Moved.